File: algorithms/javaalgo/src/test/java/programmers/level2/zip/ZipTest.java
package programmers.level2.zip;

import static org.assertj.core.api.Assertions.assertThat;
import static org.junit.jupiter.api.Assertions.assertAll;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import org.junit.jupiter.api.Test;

public class ZipTest {

    static Map<String, Integer> dict;

    @Test
    void test() {
        String msg1 = "KAKAO";
        String msg2 = "TOBEORNOTTOBEORTOBEORNOT";
        String msg3 = "ABABABABABABABAB";

        assertAll(
            () -> assertThat(solution(msg1)).isEqualTo(new int[]{11, 1, 27, 15}),
            () -> assertThat(solution(msg2)).isEqualTo(
                new int[]{20, 15, 2, 5, 15, 18, 14, 15, 20, 27, 29, 31, 36, 30, 32, 34}),
            () -> assertThat(solution(msg3)).isEqualTo(
                new int[]{1, 2, 27, 29, 28, 31, 30})
        );

    }

    public int[] solution(String msg) {
        List<Integer> answers = new ArrayList<>();
        dict = new HashMap<>();
        initDict();
        int dictIndex = 27;
        int index = 0;
        int length = msg.length();

        while (index < length) {
            if (index == length - 1) {
                answers.add(dict.get(String.valueOf(msg.charAt(index))));
                break;
            }
            String word = String.valueOf(msg.charAt(index));
            index++;
            while (index < length) {
                String nextWord = String.valueOf(msg.charAt(index));
                if (!dict.containsKey(word + nextWord)) {
                    dict.put(word + nextWord, dictIndex);
                    dictIndex++;
                    break;
                }
                word += nextWord;
                index++;
            }
            answers.add(dict.get(word));
        }
        return answers.stream().mapToInt(i -> i).toArray();
    }

    public void initDict() {
        int index = 1;
        for (int i = 'A'; i <= 'Z'; i++) {
            dict.put(String.valueOf((char) i), index++);
        }
    }
}

File: algorithms/javascriptalgo/programmers/level2/zip.test.js
test("test", () => {
    let msg1 = "KAKAO";
    let msg2 = "TOBEORNOTTOBEORTOBEORNOT";
    let msg3 = "ABABABABABABABAB";

    expect(solution(msg1)).toEqual([11, 1, 27, 15]);
    expect(solution(msg2)).toEqual([20, 15, 2, 5, 15, 18, 14, 15, 20, 27, 29, 31, 36, 30, 32, 34]);
    expect(solution(msg3)).toEqual([1, 2, 27, 29, 28, 31, 30]);
});

function solution(msg) {
    let answer = [];
    const dict = "ABCDEFGHIJKLMNOPQRSTUVWXYZ".split("")
        .reduce((acc, cur, i) => (acc[cur] = i + 1, acc), {});

    let nextWord = "";
    let wordCount = 27;
    const word = msg.split("").reduce((acc, cur) => {
        nextWord = acc + cur;
        if (dict[nextWord] === undefined) {
            dict[nextWord] = wordCount++;
        } else {
            return acc + cur;
        }
        if (dict[acc] !== undefined) {
            answer.push(dict[acc]);
        }
        return cur;
    }, "");

    answer.push(dict[word]);

    return answer;
}
